### ISSUE TITLE
RFC: Remove Nested(many=True) to keep only the List(Nested()) form

### ISSUE THREAD

**Issue by lafrech** [marshmallow-code/marshmallow#779]
See discussion in https://github.com/marshmallow-code/marshmallow/issues/493#issuecomment-380468027.

We have two ways to do the same thing. They sometimes behave differently, which generally means one of them is buggy.

IMHO, and from what I have seen from users around me, it is more natural for a new user to write `List(Nested)` than `Nested(many)`. Besides, it is consistent with `Dict(values=Nested)`.

I realize this is quite a breaking change, both from user perspective, because `Nested(many=True)` is probably the most commonly used form, and from developer perspective, because it involves a lot of code changes (but also a lot of code removal, which is nice).

There could be things that can be done with `Nested(many=True)` but not with `List(Nested)`, because in the former case, the nested Schema has access to the whole data, while in the latter, it is called on each element. Perhaps some `pre_load` edge cases might suffer from the change. Any real-life example, anyone?

Anyway, since the question was raised in a comment in another issue, I figured I'd give it more visibility.

**Comment by seansfkelley** [comment]
Is there any documentation outlining the differences between these two? I find the distinction confusing as well.

**Comment by deckar01** [comment]
I don't think there is supposed to be a difference.

**Comment by sloria** [comment]
I think deprecating `Nested(many=True)` is a good idea. Would someone be up for doing an analysis of the current differences between the two APIs?

**Comment by lafrech** [comment]
Do you mean deprecate in 2.X and remove in 3 or deprecate in 3?

Here are a few differences:


- Validation error issue with `Nested(many=True)` (https://github.com/marshmallow-code/marshmallow/issues/493#issuecomment-380468027) (`List(Nested)` does the right thing.)
- Different error message when deserializing an invalid input (not a list) (https://github.com/marshmallow-code/marshmallow/pull/954#issuecomment-421916563) (`List(Nested)` does the right thing.)
- `List(Nested)` prevents pre/post processors in nested schema to work on the whole data collection] (https://github.com/marshmallow-code/marshmallow/issues/1801)

------------------------------

- [Partial data if an item fails validation](https://github.com/marshmallow-code/marshmallow/issues/766) [#766] (Fixed in #1052)
- [`List` is not propagating the context between a nested field and the parent schema](https://github.com/marshmallow-code/marshmallow/issues/922) [#922] (Fixed)
- [`only` argument inconsistent between Nested(S, many=True) and List(Nested(S))](https://github.com/marshmallow-code/marshmallow/issues/946) [#946] (Fixed in  #1229)
- `partial` not propagated to `List(Nested)` (Fixed in #1230)

**Comment by sloria** [comment]
@lafrech I was thinking of deprecating in 3. As you pointed out, `Nested(many=True)` is the probably the more common usage, and removing it would break many users' code.

**Comment by sloria** [comment]
How would a self-referencing schema work with this? In other words, how would we express 

```python
friends = Nested('self', many=True, exclude=('friends',))
```

**Comment by deckar01** [comment]
I would expect `friends = List(Nested('self', exclude=('friends',)))` to work the same way, but it currently errors. Using a reference to the schema with python 3 works, so I suspect this is just an edge case that hasn't been tested.

**Comment by ex5** [comment]
> Perhaps some pre_load edge cases might suffer from the change. Any real-life example, anyone?

I've got production code that relies on the fact that a nested field with `many=True` and `@pre_dump(pass_many=True)` on its schema will execute this `pre_dump` **only once**.

The benefit is making sure expensive code that is needed to be executed before nested fields can be dumped is not repeated for each element in the `fields.List`.

I don't really see this as an edge case though, pretty sure doing something to the whole set of related objects before dumping/loading them is a common pattern.

**Comment by deckar01** [comment]
Currently a list of nested fields does not propagate the schema context correctly, but nested many does. See #424 and #922.

**Comment by lafrech** [comment]
Yep, I just updated [the list above](https://github.com/marshmallow-code/marshmallow/issues/779#issuecomment-393283257).

**Comment by deckar01** [comment]
I saw the ref on the issue, but I couldn't figure out where it came from. Thanks.

**Comment by rooterkyberian** [comment]
@anka-sirota yes it seems like a common use case, but it very well could be done on parent schema or by subclassing `List` and doing custom `pre_dump` there. All that is possible without handling a separate `many` case, which I would I agree with @lafrech, just source of inconsistent in the library.

Looking at bigger picture, that this issue already have been here for couple months now, and there still few bugs before `many=True` can be complete eliminated from the library, I would suggest just calling it `deprecated` in marshmallow 3 release.
Such move will:
* prevent further delay of marshmallow 3
* make it so that there is more interest in `List(Nested)` and that it works consistently with `Nested(many=True)`

**Comment by rooterkyberian** [comment]
@lafrech another difference to be added to the list https://github.com/marshmallow-code/marshmallow/issues/946

**Comment by lafrech** [comment]
@rooterkyberian, I was just searching for this thread to reference your issue.

**Comment by lafrech** [comment]
All differences listed above are either fixed or on their way to be fixed. Once this is done, we should be able to deprecate `Nested(many=True)`.

**Comment by bartelemi** [comment]
There is also a difference between `Nested(Schema, many=True)` and `List(Nested(Schema))` when loading partial data. Below tests are showing the difference in behaviour when using both methods:

```py
from marshmallow import Schema
from marshmallow.fields import List, Nested, Str


class Employee(Schema):
    name = Str(required=True)
    title = Str(required=True)


def test_nested_partial_load_with_list():
    class Company(Schema):
        employees = List(Nested(Employee))

    payload = {
        'employees': [{
            'name': 'Arthur',
        }],
    }
    # Partial loading with list type shouldn't generate any errors.
    result = Company().load(payload, partial=True)
    assert result['employees'][0]['name'] == 'Arthur'


def test_nested_partial_load_with_nested_many():
    class Company(Schema):
        employees = Nested(Employee, many=True)

    payload = {
        'employees': [{
            'name': 'Arthur',
        }],
    }
    # Partial loading with Nested(many=True) type shouldn't generate any errors.
    result = Company().load(payload, partial=True)
    assert result['employees'][0]['name'] == 'Arthur'
```

Partial loading in nested schemas was introduced in version 3.0.0rc1 in this PR: #849. Problem is that when a field is of type `List`, it is not getting the same `kwargs` as `Nested` so it can't ignore missing data. I think it was even mentioned in referenced discussion.

**Comment by lafrech** [comment]
Thans @0xpsilocybe, I think this is the object of [@sloria's comment](https://github.com/marshmallow-code/marshmallow/pull/1066#issuecomment-449638943) in #1066. I shall update the PR to propagate all needed `kwargs` to Nested.

**Comment by timc13** [comment]
What about the `Pluck(many=True)` use case ?

**Comment by deckar01** [comment]
Removing `many` would allow a few lines of explicit `many` handling to be removed from `Pluck`. The rest of the changes would automatically be inherited from `Nested`.

**Comment by sloria** [comment]
@lafrech Is there anything else to be addressed before we can deprecate `Nested(many=True)`? I think we'll want to investigate https://github.com/marshmallow-code/marshmallow/issues/779#issuecomment-396354987 .

In any case, I don't think we should block on this for releasing 3.0. We can always deprecate within the 3.x line.

**Comment by fuhrysteve** [comment]
@sloria :+1: on depreciating stuff rather than stacking the deck with more BC breaks. I'd much rather have 8 releases with 8 breaking changes than 8 breaking changes in one release...

I don't know about anyone else, but we've got hundreds - maybe even a thousand schemas to migrate. I'd rather take that in strides!

**Comment by lafrech** [comment]
@fuhrysteve, about this, please see the conversation in https://github.com/marshmallow-code/marshmallow/issues/928.

**Comment by sloria** [comment]
Changed the milestone on this since there are still further considerations to discuss (deprecating Nested(many=True)), but I don't think there's any action items to complete before 3.0 final.

**Comment by lafrech** [comment]
Indeed, deprecation can be done in 3.x so it doesn't have to hold 3.0.

I sent https://github.com/marshmallow-code/marshmallow/pull/1290 with a fix for that last `List(Nested(...))` glitch we identified above (https://github.com/marshmallow-code/marshmallow/issues/779#issuecomment-396354987).

**Comment by sloria** [comment]
Before we officially recommend `List(Nested(Schema))`, it would be good to do some benchmarking and profiling. A quick-n-dirty benchmark shows that `List(Nested(Schema))` is ~35% slower than `Nested(Schema, many=True)`: https://gist.github.com/sloria/2e0b141b3d05fe0dafbdeafd294ebae8

**Comment by deckar01** [comment]
I might try to dig into a perf dump and see where the time is going. That seems like a lot of time to just be caused by an extra layer of indirection. I wonder if `Nested` isn't doing something it should be or if `List` is doing something it doesn't need.

**Comment by deckar01** [comment]
I ran [the benchmarks above](https://gist.github.com/sloria/2e0b141b3d05fe0dafbdeafd294ebae8) using python 3.7.2 against the `dev` branch. I had to make a copy of `Schema` to get sane `cProfile` output, because recursion merges calls into a single total and hides the problem.

**Nested(many)**

<img width="1034" alt="Screen Shot 2019-07-16 at 11 03 23 AM" src="https://user-images.githubusercontent.com/3108007/61312084-52200b00-a7bd-11e9-9361-86864446d6f0.png">
<img width="961" alt="Screen Shot 2019-07-16 at 11 04 30 AM" src="https://user-images.githubusercontent.com/3108007/61312058-4a606680-a7bd-11e9-9db1-483b9ecea58d.png">

**List(Nested)**

<img width="1035" alt="Screen Shot 2019-07-16 at 11 03 42 AM" src="https://user-images.githubusercontent.com/3108007/61312061-4a606680-a7bd-11e9-9530-311d00dff606.png">
<img width="973" alt="Screen Shot 2019-07-16 at 11 04 44 AM" src="https://user-images.githubusercontent.com/3108007/61312057-4a606680-a7bd-11e9-8f3c-eee9aaf952cd.png">

---

It looks like the slowdown is caused by `Nested(many)` calling `schema.dump(many)` once and getting to loop on `schema._serialize()`, but `List(Nested)` having to loop on `schema.dump()`.

About 20% of that slowdown is caused by calls to `_has_processors()` which is cacheable and `ErrorStore()` which could possibly be removed or refactored as a shared instance. Optimizing these would also make `schema.dump()` calls slightly faster in general.

The rest of the slow down seems to be memory allocation and conditional evaluation in `dump()`. The easiest way to fix that is probably to probe the list's inner field type and allow it to hit the optimized schema path once via `Nested._serialize(many)`.

**Comment by deckar01** [comment]
#1304 makes `dump()` ~18% faster for both paths. #1306 makes `dump()` ~21% faster for `List(Nested)`, which makes the two paths run in effectively the same amount of time (< 1%).

**Comment by lafrech** [comment]
Great.

Is there the same kind of gap between `Nested(many)` and `List(Nested)` when loading?

**Comment by deckar01** [comment]
Ya, `load()`has the same ~35% slowdown when comparing `List(Nested)` to `Nested(many)`. The same strategy should work there.

**Edit:** I added the fix for `load()` in #1306 and confirmed that it makes them run in the same time without slowing down `Nested(many)`.

**Comment by jtrakk** [comment]
Does it make sense to remove `Integer(many=True)` and the `many` argument on other fields too?

**Comment by sloria** [comment]
@jtrakk `many` isn't an argument to other fields.

**Comment by jtrakk** [comment]
I see. FWIW, I've seen people (including myself) pass `Integer(many=True)` and been confused when it doesn't work. I know [`attr.ib()`](http://www.attrs.org/en/stable/api.html#attr.ib) uses the `metadata=` namespace instead of taking `**metadata`, to avoid that problem, at the cost of a little extra verbosity. I'm not sure how I feel about the tradeoff, though I lean toward the explicit namespace.

**Comment by sloria** [comment]
Yeah, it's a known issue. The reasons we use **kwargs instead of e.g. `metadata=` are mostly historical. The original decision was that storing kwargs 1) was more concise and 2) saved us from having to come up with an appropriate name... "metadata" didn't seem right because there are use cases where the things your storing aren't really metadata. At this point, it's not worth breaking the API.

Not the best reasons, but I think it's not terrible. We've discussed adding a [whitelist of metadata keys](https://github.com/marshmallow-code/marshmallow/issues/683#issuecomment-385113845) in the past, but we decided it wasn't worth the added API surface.

**Comment by jtrakk** [comment]
> At this point, it's not worth breaking the API.

I suppose so.

That being said, 3.0.0 would be a good time to do it, and it [wouldn't be hard](https://instagram-engineering.com/static-analysis-at-scale-an-instagram-story-8f498ab71a0c) to write an automatic converter using [libCST](https://libcst.readthedocs.io/en/latest/) or similar that'd move all the current `**metadata` calls into `metadata=dict(...)`.

**Comment by sloria** [comment]
@jtrakk It's too late for 3.0 (planning to release the final today 😬 ), but I've opened https://github.com/marshmallow-code/marshmallow/issues/1350 for further discussion.

**Comment by sloria** [comment]
With this change and the `lambda` change proposed [here](https://github.com/marshmallow-code/marshmallow/issues/1146#issuecomment-529157000), we can also probably deprecate the `only`, `exclude`, and `unknown` parameters of `Nested`.

```python
# before
albums = fields.Nested(AlbumSchema, many=True, only=('id', 'title', ))
# after (possible now)
albums = field.List(fields.Nested(AlbumSchema(only=('id', 'title'))))

# before
artist = fields.Nested('ArtistSchema', only=('id', ))
# after (proposed)
artist = fields.Nested(lambda: ArtistSchema(only=('id',)))
```

**Comment by deckar01** [comment]
While working on #1498, @Meallia uncovered that `Nested(many)` does not support enabling `allow_none` for list items. `Schema.load()` has no concept of `allow_none`, so it will always implicitly be true.

**Comment by ThiefMaster** [comment]
marshmallow-code/marshmallow#1801 may be related to this topic.

But after looking at the code of `List` it seems like it probably wouldn't be particularly hard to make it use many on the underlying schema instead of calling it in a loop over the items...

**Comment by mwgamble** [comment]
Maybe the best solution is to document the differences and otherwise leave it be. It's easy to imagine there are many codebases around the world that depend on one of the two behaviours, given they aren't identical.

**Comment by lafrech** [comment]
`List(Nested())` happens to be faster according to #2872.

**Comment by lafrech** [comment]
OK, so the only thing that can currently only be achieved with `Nested(many)` seems to be pre/post processing collections.

The current implementation allows users to optionally allow passing collections, in which case the processor receives a boolean argument indicating whether or not it is fed a collection or a single item. It is then up to the processor to manage both cases. [The docs](https://marshmallow.readthedocs.io/en/latest/extending/pre_and_post_processing_methods.html#passing-many) provide the example (enveloping data in the payload).

While this may allow for a little factorization, it just occurred to me that things might be about as simple if the user declared one function for each case (single/many). They might even just need one case in practice, who knows?

My point is that the single and many cases are two different things that could be addressed by different functions. We could drop the `pass_collection` argument and have a separate decorator for the many case, so that `List(Nested)` calls `@pre/post_load_many` with collections. In fact, we may even be able to achieve the same goal while keeping the current API.

Bottom line: `List(Nested)` calls collection-level pre/post processors passing a collection, but keeps (de)serializing items one by one using the `Nested` schema that calls item-level processors.

(Looks like the docs are outdated. They still mention `pass_many` which was renamed into `pass_collection`.)
